src/config.rs

```diff
@@ -1,33 +1,140 @@
+use std::{
+    collections::HashMap,
+    fs::File,
+    path::{Path, PathBuf},
+    process,
+};
+
+/// Restart conditions for a service.
+#[derive(Serialize, Deserialize, Debug)]
+#[serde(rename_all = "lowercase")]
+enum Restart {
+    Never,
+    Always,
+    Unexpected,
+}
 
-#[derive(Debug)]
 /// Configuration structure used to run a task.
+#[derive(Serialize, Deserialize, Debug)]
+#[serde(deny_unknown_fields)]
+#[serde(default)]
 pub struct Config {
     /// Command to run.
     cmd: String,
     /// Number of processors this task should run with.
     // u16 is fine because we are not expecting a machine to run more
     // than 2^16 proc at any single time.
     numprocs: u16,
+    /// Program mask running under
+    umask: i32,
     /// Working directory of the task.
     workingdir: PathBuf,
+    /// Service starting at load time.
+    autostart: bool,
+    /// Service restarting when finished.
+    autorestart: Restart,
+    /// Autorestart callback trigger value.
+    exitcodes: Vec<i32>,
+    /// Restart attempts.
+    startretries: i32,
+    /// Delay before start.
+    starttime: i32,
+    /// Signal received by supervisord trigger service termination.
+    stopsignal: String,
+    /// Running period.
+    stoptime: i32,
+    /// Stdout logfile path.
+    stdout: PathBuf,
+    /// Stderr logfile path.
+    stderr: PathBuf,
+    /// Environment variable service is running into.
+    env: HashMap<String, String>,
 }
 
 impl Default for Config {
     fn default() -> Self {
+        let bin = "ls";
+        let home = dirs::home_dir();
+        if home.is_none() {
+            eprintln!("Impossible to get user home directory!");
+            process::exit(1);
+        }
+        let home = home.unwrap();
+
+        #[cfg(debug_assertions)]
+        println!("{:?}", home);
+
         Self {
-            cmd: "ls".into(),
+            cmd: bin.into(),
             numprocs: 1,
-            workingdir: PathBuf::from("/tmp"),
+            umask: 0o027,
+            workingdir: home.clone(),
+            autostart: true,
+            autorestart: Restart::Never,
+            exitcodes: {
+                let mut v = Vec::new();
+                v.push(0);
+                v
+            },
+            startretries: 0,
+            starttime: 0,
+            stopsignal: "SIGTERM".to_string(),
+            stoptime: 0,
+            stdout: home.join(bin.to_string() + ".stdout"), // Would be lovely to replace "bin" by "cmd" from the struct itself
+            stderr: home.join(bin.to_string() + ".stderr"), // Same
+            env: HashMap::new(),
         }
     }
 }
 
-impl std::convert::TryFrom<&Path> for Config {
-    type Error = std::io::Error;
+/// # Errors
+///
+/// Will return `Err` if `file` does not exist or the user does not have
+/// permission to read it or yaml format is not respected.
+pub fn parse(
+    file_name: &Path,
+) -> Result<HashMap<String, HashMap<String, Config>>, serde_yaml::Error> {
```

Comment:
Why is `parse` returning this type, and not a `Config` file?

Reply:
Because of the format of the config file:
```
programs:
         nginx:
                cmd: "ls -lF"
                ...
         vogsphere:
                cmd: "/sbin/vogsphered"
                ...
```
(config.yaml)

<HashMap<"programs", HashMap<"nginx", Config {cmd: "ls -lF"}> ...>
 
What is the best way to handle this ? Have you encounter this issue before ?
I could possibly change the `Config` struct for it to hold the HashMap then TryFrom could work ?

Moreover, there would be additional sections for daemon to be customized.
```
programs:
         nginx:
                cmd: "ls -lF"
                ...
         vogsphere:
                cmd: "/sbin/vogsphered"
                ...
daemon:
        ip: "192.168.1.25"
        port: "447"
        user: "root"
        
```
(another_config.yaml)

Reply:
We generally tend to type things. So if you had something like `HashMap<String, HashMap<String, Config>>` you could do something like:
```rust
type TaskMasterConfig: HashMap<String, HashMap<String, Config>>;
```

However, I think we can do much better
```rust
struct TaskMasterConfig {
  programs: HashMap<String, ProgramConfig>,
  daemon: DaemonConfig,
}

struct DaemonConfig {
  ip: IpAddr,
  port: Port,
  user: String,
}

struct ProgramConfig {
  cmd: String,
  something_that_s_optional: Option<String>,
}
```

What do you think?